### ISSUE TITLE
IP.block set by a variety of rules in 910 and 913, leads to unconditional block in 949100

### ISSUE THREAD

**Issue by dune73** [SpiderLabs/owasp-modsecurity-crs#514]
949100 block unconditionally if ip.block is set. ip.block is set by various rules in 910 and 913 regardless of scoring or blocking policy or any other setting in crs-setup.conf.

We need to discuss this blocking-on default policy for IP repudiation. And we need to describe an opt-out mechanism. At least. In fact I think we need far more documentation for this.

Personally I would prefer if we had a default `repudiation-blocking=off` policy by default. People will raise their anomaly thresholds to test out the CRS and then surprisingly get arbitrary blocks on bad behaviour of individual clients. This gives a scary feeling.

**Comment by lifeforms** [comment]
I found this feature very puzzling too.

Having clients getting blocked unexpectedly is the worst. Out of desperation you restart the web server process, but the clients are being kept blocked until you delete the collection databases.

It's also pretty hard to reason about the scenarios that contribute to a block. Why not just block after a certain number of CRS blockings / error responses from the back-end. We actually do this on the application level now. The additional 403s (not blocked by CRS but by the ip block) are a great way to get audit logs on unknown attack types, since an attacker often tries many things until one gets through.

When you really WANT the blocking feature, `ip.block` turns out to be actually misnamed, since it actually blocks not the IP but an [User-Agent, IP] tuple, which already puzzled me in #442. This behavior also seems ripe for a configuration flag.

So, at least we should have much better documentation about it, but I think that some configuration knobs are absolutely in order, like (1) enable/disable persistent blocking, (2) choose the reasons for a persistent block, (3) the scope for a block (UA or full IP). However, this could be a serious project. We could start with the enable/disable flag but keep in mind that it might be part of a larger set of options in the future.

For now, I think I agree with you that it's probably better if the default setting is off.

I've also had a serious collections performance issue in 2.9.1 a few weeks back (requests dropping from 250 req/s to 50 req/s on my testing server due to SOME entry in collections - this persisted for days until I debugged it so normal expiry didn't fix it). Which would be another argument for NOT using collections unless the user explicitly wants it.

**Comment by dune73** [comment]
Thanks for adding more food for thought.

So our common ground would be:
 - Introduce config flag for CRS3.0
 - New config flag is off for CRS3.0
 - rework the whole IP/UA Blocking stuff for CRS3.1

Correct?

@csanders-git : What do you think about this?

Looking over my FPs the biggest number (outside of 920320 which we disabled) is 949100 and it is a 403. :scream:

**Comment by dune73** [comment]
Here is a brief overview on the use of `IP.block`:

```
READS:
		910000	Skip based on ip.block
		949100	BLOCK: Request Denied by IP Reputation Enforcement 

SETS:
		910100	High risk country
		910110	IP Blacklist (commented out) Trustwave SpiderLabs
		910150	HTTP Blacklist match for search engine IP based on tx.block_search_ip
		910160	HTTP Blacklist match for spammer IP based on tx.block_spammer_ip
		910170	HTTP Blacklist match for suspicious IP based on tx.block_suspicious_ip
		910180	HTTP Blacklist match for harvester IP based on tx.block_harvester_ip
		913100	Found User-Agent associated with security scanner
		913110	Found request header associated with security scanner
		913120	Found request filename/argument associated with security scanner
		913101	Found User-Agent associated with scripting/generic HTTP client
		913102	Found User-Agent associated with web crawler/bot

```

**Comment by dune73** [comment]
@lifeforms mentioned the performance issues of the IP collection. It would make sense to couple the initialisation of the said collection with the setting of our new blocking switch. However, the ip collection is also used in the DOS protection and I think it is too late to interfere with so many rules for an RC2. I will thus leave the creation of the IP collection aside and make the most simple change which allows us to disable IP blocking based on repudiation by default.

**Comment by dune73** [comment]
Closing this in favor of PR #580.
